✨ feat(HAT-142): Spring Security Setting

IToriginal opened the pull request:
## Motivation:

- Spring Security 설정

## Modifications:

- 회원 도메인 설정을 통해 Member의 Table 생성을 위해서 Repository Test를 진행함.
- Spring Security 적용을 위해 의존성 설정을 하고 Security의 기본 사용자 변수를 앞서 만든 Member Entity에 맞춰서 설정.

## Result:

- Repository Test는 기본적인 `CRUD` 기능만 수행했음
    - 회원 정보 저장
    - 회원 정보 수정
    - 회원 정보 삭제 (Hard Delete로 구현함. 계획은 Soft delete 이므로 서비스로직의 테스트에 구현 예정)
    - 회원 정보 조회
- 사용자의 아이디(loginId)와 패스워드(loginPassword)를 이용해 Access Token과 Refresh Token을 얻기 위한 초기 단계를 구현하였음.
- 사용자의 아이디와 패스워드를 이용해서 JWT문자열을 생성하는 기능을 수행하기 위해서 사용자가 전달하는 값을 알아 낼 수 있어야 하므로 POST 방식의 JSON 문자열을 이용하기 위해 문자열을 JSON으로 처리하는 설정을 구현함

IToriginal committed:
✨ feat(HAT-93): 회원 Entity, Repository 생성 및 Test Code 작성

아래의 Repository Test를 수행하였음
1. 회원 정보 저장
2. 회원 정보 수정
3. 회원 정보 삭제
4. 회원 정보 조회

IToriginal committed:
✨ feat(HAT-103): Spring Security 사용자 처리 설정

IToriginal committed:
✨  feat(HAT-94): Token 인증을 위한 Security Filter 적용 설정

IToriginal committed:
✨  feat(HAT-143): Security LoginFilter JSON 처리

IToriginal committed:
💡  docs: 주석 처리 수정

IToriginal committed:
💡  docs: lint code 수정

IToriginal committed:
🔀  ci: 브랜치 병합

IToriginal committed:
Merge branch 'main' into feat/HAT-142-Security-Setting

IToriginal committed:
🔥 fix: check style 어노테이션 자동 추가 구문 삭제

youngsoosoo reviewed (approved):
수고하셨습니다!!

jooh9992 commented on member-domain-rds/src/test/java/io/howstheairtoday/memberdomainrds/repository/MemberRepositoryTest.java:
```diff
@@ -0,0 +1,150 @@
+    // }
+
+    @Test
+    @DisplayName("회원 정보 저장")
```
DisplayName 어노테이션은 최상단에 넣기로 했습니다. 수정 부탁 드려요!

IToriginal replied:
수정 완료했습니다!👍🏻

youngsoosoo commented on member-app-external-api/src/main/java/io/howstheairtoday/appmemberexternalapi/config/RootConfig.java:
```diff
@@ -0,0 +1,23 @@
+ * ModelMapper 사용을 위한 설정
+ */
+@Configuration
+public class RootConfig {
```
ModelMapper를 왜 사용했나요?

IToriginal replied:
ModelMapper 라이브러리는 Java 객체간의 매핑을 쉡게 처리할 수 있게 하는 라이브러리 입니다.

ModelMapper를 사용하지 않으면 Member Entity와 DTO 간의 변환에 일일히 매핑해줘야 하는데 ModelMapper를 등록하면 번거로운 작업 없이 자동으로 변환해 줍니다. 
예를 들면 dto -> Entity 이러한 코드가 작성되지 않아도 된다? 라고 이해하였습니다.

youngsoosoo replied:
좋은 정보 얻어갑니다!

jhva commented on member-app-external-api/src/main/java/io/howstheairtoday/appmemberexternalapi/config/CustomSecurityConfig.java:
```diff
@@ -0,0 +1,80 @@
+        httpSecurity.addFilterBefore(memberLoginFilter, UsernamePasswordAuthenticationFilter.class);
+
+        httpSecurity.csrf().disable();
+        httpSecurity.sessionManagement().sessionCreationPolicy(SessionCreationPolicy.STATELESS);
```
antMatcher 넣어주셔야할거같아요!

IToriginal replied:
보안 URL 설정은 뒤에 하자 생각했었는데 지금 상황에서 처리해야 할 문제인 것 같네요!
해당 부분은 다시 추가해 두겠습니다.

> 추가 반영 부분
// post 기능에 인증된 회원만 접속 가능토록 설정
httpSecurity.authorizeRequests().requestMatchers("/api/v1/post/**").authenticated().anyRequest().permitAll();

antMatchers()는 Spring Boot 3 프로젝트에서 사용 할 수 없다는 stack over flow의 도움을 받아 requestMatchers를 사용했습니다.

IToriginal replied:
현재 이 인증 기능을 추가해두었지만 회원 기능이 완료되면 다시 활성화한 뒤에 테스트를 해볼 예정이라 추가 했던 부분은 다시 주석처리로 비활성화 하였습니다. 참고 부탁합니다.

IToriginal committed:
✨ feat: 특정 URL에 인증된 회원만 접속 가능한 설정 추가

IToriginal commented on member-app-external-api/src/main/java/io/howstheairtoday/appmemberexternalapi/service/MemberDetailsService.java:
```diff
@@ -0,0 +1,46 @@
+
+    private final MemberRepository memberRepository;
+
+    @SuppressWarnings("checkstyle:AbbreviationAsWordInName")
```
감사합니다. 수정 완료 했습니다.!! 👍🏻

IToriginal committed:
💡 docs: DisplayName 어노테이션 위치 수정

jooh9992 reviewed (approved):
다음 PR이 기대됩니다!

IToriginal committed:
💡 docs: URL 인증 기능 주석 처리

회원 생성 기능이 미완료 상태이기 때문에 현재 기능을 추가하면 다른 기능에 영향을 끼칠수 있으므로 주석처리하여 회원 생성 기능을 완료한 뒤에 기능을 활성화할 예정

Hyyena commented on member-app-external-api/src/main/java/io/howstheairtoday/appmemberexternalapi/service/MemberDetailsService.java:
```diff
@@ -0,0 +1,46 @@
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+
+        Optional<Member> result = memberRepository.findByLoginId(username);
```
`findByLoginId(username)` 에서 `username`은 로그인 아이디를 뜻하는 건가요?

`memberRepository`에서 정의하신 `findByLoginId`가 필요로 하는 매개변수는 `loginId`로 적혀있어서 여쭤봅니다!

Hyyena commented on member-app-external-api/src/main/java/io/howstheairtoday/appmemberexternalapi/config/CustomSecurityConfig.java:
```diff
@@ -0,0 +1,80 @@
+
+        // MemberLoginFilter
+        // Spring Security에서 username과 password를 처리하는 UsernamePasswordAuthenticationFilter의 앞쪽에서 동작하도록 설정
+        MemberLoginFilter memberLoginFilter = new MemberLoginFilter("/generateToken");
```
이 부분이 로그인 Request 시 해당 요청 URL을 필터로 인터셉트하여 Spring Security가 처리하는 부분인가요?

그렇다면 로그인 요청 URL인 `api/v1/auth/signin`이어야 하는 것 아닌가요?

```java
MemberLoginFilter memberLoginFilter = new MemberLoginFilter("/api/*/auth/signin");
```

IToriginal replied:
해당 부분은 URL을 판단하는 구문이 아니라, 인증할 토큰이 있냐 없냐를 처리한다고 생각하시면 될 듯합니다.
말씀해주신 URL은 아직 기능을 처리하지는 않았습니다. Post를 예로 작성해두긴 했는데 해당 구문은 회원기능이 완료되지않는다면 Post URL에서 인증이 되지 않아 개발에 어려움을 주게 됩니다. 그래서 요청 URL의 경우는 회원 가입 기능을 완료한 뒤에 이루어 질 것입니다.

IToriginal replied:
다시 생각해보니 해당 구문은 현우님 comment 내용이 맞는 거 같습니다. 
잘못된 comment로 혼선을 드렸네요🥲 예제를 제대로 이해하지 못했던것 같습니다. 감사합니다. 
해당 내용은 인지한 상태로 조금 더 확인 한 후에 다음 작업이나 로그인을 구현할때 변경 하도록 하겠습니다👍

Hyyena reviewed (commented):
> 회원 정보 삭제 (Hard Delete로 구현함. 계획은 Soft delete 이므로 서비스로직의 테스트에 구현 예정)

도메인 로직을 서비스 로직의 테스트에서 구현하시려는 이유를 알 수 있을까요?

IToriginal commented on member-domain-rds/src/main/java/io/howstheairtoday/memberdomainrds/repository/MemberRepository.java:
```diff
@@ -0,0 +1,17 @@
+
+import io.howstheairtoday.memberdomainrds.entity.Member;
+
+@Repository
```
DataJpaTest 어노테이션 사용시 Bean을 사용하기 위해서는 해당 어노테이션이 자동으로 입력됩니다.
어노테이션을 삭제하는 순간 테스트 구문에서 MemberRepository를 읽지 못하는 에러가 발생하게 되는 현상이 존재 했습니다.
따라서 해당 어노테이션은 DataJpaTest를 사용하게 된다면, 불가피하게 등록해야하는 어노테이션이 됩니다🥲

Hyyena commented on member-domain-rds/src/main/java/io/howstheairtoday/memberdomainrds/entity/Member.java:
```diff
@@ -0,0 +1,71 @@
+    @Id
+    @GeneratedValue(generator = "uuid2")
+    @GenericGenerator(name = "uuid2", strategy = "org.hibernate.id.UUIDGenerator")
+    @JdbcTypeCode(SqlTypes.VARCHAR)
```
[Confluence](https://kakao-4.atlassian.net/wiki/spaces/SD/pages/8028182/DB+UUID) 확인 한 번 부탁드립니다!

![image](https://user-images.githubusercontent.com/70932170/227928650-dcaf9960-d0d1-4c25-a222-26b0c2a08b3a.png)

IToriginal commented:
> 

답변?

IToriginal commented on member-domain-rds/src/main/java/io/howstheairtoday/memberdomainrds/entity/Member.java:
```diff
@@ -0,0 +1,74 @@
+@ToString
+public class Member extends BaseTimeEntity {
+
+    @Id
+    @GeneratedValue(generator = "uuid2")
+    @GenericGenerator(name = "uuid2", strategy = "org.hibernate.id.UUIDGenerator")
+    @JdbcTypeCode(SqlTypes.VARCHAR)
+    private UUID memberId;
+    @Column(nullable = false, length = 40)
+    private String loginId; // 아이디
+    @Column(nullable = false, length = 128)
+    private String loginPassword;
+    @Column(nullable = false, length = 50)
+    private String email;
+    @Column(nullable = false, length = 40)
+    private String nickname;
+    @Column(length = 128)
+    private String memberProfileImage;
+    @Enumerated(EnumType.STRING)
+    private LoginType loginType;
+    @Enumerated(EnumType.STRING)
+    @Column(nullable = false)
+    private LoginRole loginRole;
+    @Column(length = 255)
+    private String token; // refreshToken
```
참고 하겠습니다. 감사합니다.👍🏻

IToriginal replied:
![image](https://user-images.githubusercontent.com/117193889/227934450-6ba14fc3-a08b-41c6-8e15-73ddb8b540a0.png)
방금 수정을 하려고 보는데 저희가 지정한 체크 스타일에서 한번 걸리긴 하는 거 같은데 무시하고 줄바꿈을 시행할 지 의견이 필요할 꺼 같네요!

IToriginal commented on member-domain-rds/src/main/java/io/howstheairtoday/memberdomainrds/entity/Member.java:
```diff
@@ -0,0 +1,74 @@
+    @Column(nullable = false, length = 40)
+    private String loginId; // 아이디
+    @Column(nullable = false, length = 128)
+    private String loginPassword;
```
안그래도 해당 문제에 대해서 공유하려 했는데, Security를 사용하게 되면 Security에서의 ID 변수를 username, PW의 변수를 password로 사용합니다. 그래서 저희가 작성한 ERD 회원 테이블의 password와 security에서의 정해진 변수 password와 간혹 구분이 어려운 경험이 생겼어서 변경하였습니다.  팀원들의 동의를 구하여야하나 선조치에 오해의 소지를 불렀네요..🫠
팀원들의 의견이 중요하므로, loginPassword로 사용하는 것이 괜찮다면 ERD 회원 테이블에 명시된 값을 수정하겠습니다.

IToriginal commented on member-domain-rds/src/main/java/io/howstheairtoday/memberdomainrds/entity/Member.java:
```diff
@@ -0,0 +1,74 @@
+    private LoginType loginType;
+    @Enumerated(EnumType.STRING)
+    @Column(nullable = false)
+    private LoginRole loginRole;
```
해당 내용도 위와 비슷한 이유로 변경했습니다. 상관 없다면 ERD Table을 수정해두겠습니다!

Hyyena replied:
답변 감사합니다:)
데일리 스크럼을 통해 공유해주시면 좋을 것 같습니다!

IToriginal commented on member-domain-rds/src/main/java/io/howstheairtoday/memberdomainrds/entity/Member.java:
```diff
@@ -0,0 +1,74 @@
+    @Column(nullable = false)
+    private LoginRole loginRole;
+    @Column(length = 255)
+    private String token; // refreshToken
```
어떤 token인지 인지못한 상태서 작성했었던 터라 수정하지 못했네요. 해당 내용을 명시적으로 refreshToken으로 수정하여 두겠습니다.

IToriginal commented on member-domain-rds/src/test/java/io/howstheairtoday/memberdomainrds/repository/MemberRepositoryTest.java:
```diff
@@ -0,0 +1,145 @@
+    @DisplayName("회원 정보 저장")
+    @Test
+    void saveMemberInfo() {
+
+        // given
+        memberRepository.save(Member.builder()
+            .loginId("hat_id")
+            .loginPassword("test123")
+            .email("kcshat@gmail.com")
+            .nickname("hat")
+            .memberProfileImage("default.jpg")
+            .loginType(LoginType.LOCAL)
+            .loginRole(LoginRole.ROLE_USER)
+            .token("TEST.REFRESH.TOKEN")
+            .build());
+
+        // when
+        List<Member> memberList = memberRepository.findAll();
+
+        // then
+        Member member = memberList.get(0);
+        assertThat(member.getLoginId()).isEqualTo("hat_id");
+    }
+
+    @DisplayName("회원 정보 삭제")
+    @Test
+    public void deleteMemberInfo() {
```
이번 코드 리뷰에 신경 써주셔서 감사합니다.👍🏻  해당 Comment 내용을 참고하여 수정을 마친 후 Slack을 통해 공유 드리겠습니다.

Test 코드에 대한 저의 생각 🛠️:
Test 코드를 잘 작성하는 것도 중요하지만, Repository 의 역할에 대해서 필요한 부분을 테스트 하는 것은 만족한다고 생각합니다.
(테스트 코드는 빠르게 작성후 확인하는 것이 더 중요하다고 저는 생각합니다 🫠.)

IToriginal replied:
주석을 이용한 BDD 방식으로 테스트 코드를 수정했습니다. 
크게 다른 부분은 없고, test에 대한 테스트값 변경, 회원 정보 저장 정도 수정했습니다.
<img width="278" alt="image" src="https://user-images.githubusercontent.com/117193889/228079504-f8e68d25-f9af-41a7-a043-b02c677f6ec6.png">

Hyyena replied:
고생하셨습니다:)

IToriginal commented on member-app-external-api/src/main/java/io/howstheairtoday/appmemberexternalapi/config/CustomSecurityConfig.java:
```diff
@@ -0,0 +1,80 @@
+        log.info("🛠️ configure -------------------- 🛠️");
+
+        // filter.MemberLoginFilter - AuthenticationManager 설정
+        AuthenticationManagerBuilder authenticationManagerBuilder =
```
configure 로 따로 선언하는 방법을 말씀하시는 거지요? 그 방법으로도 해보려고 했는데 우선 자료들이 Spring Boot 3.0 이하의 버전이 많아서 deprecated 된 부분이 많더라고요. 그래서 우선은 저희가 배운 방법으로 시도해서 실행중입니다.

IToriginal committed:
♻️ refactor: Test Code 수정

IToriginal commented:
@Hyyena Q: 도메인 로직을 서비스 로직의 테스트에서 구현하시려는 이유를 알 수 있을까요?
이 부분에 대해서 바로 아래에 댓글이 달리지 않아서 댓글로 답변 드리겠습니다.
QueryDSL의 도입으로 softDelete 작성법을 아직 알지 못해서 실제 구현 로직을 담당할 서비스 로직에서 구현을 할 예정입니다.
그렇다고 Repository Test를 중요치 않게 보진 않지만, SoftDelete 구현은 Service Test에서 구현해도 될 문제라고 판단했습니다.

IToriginal committed:
♻️ refactor: MemberLoginFilter URL Code Refactor

Hyyena commented:
> @Hyyena Q: 도메인 로직을 서비스 로직의 테스트에서 구현하시려는 이유를 알 수 있을까요?
> 이 부분에 대해서 바로 아래에 댓글이 달리지 않아서 댓글로 답변 드리겠습니다.
> QueryDSL의 도입으로 softDelete 작성법을 아직 알지 못해서 실제 구현 로직을 담당할 서비스 로직에서 구현을 할 예정입니다.
> 그렇다고 Repository Test를 중요치 않게 보진 않지만, SoftDelete 구현은 Service Test에서 구현해도 될 문제라고 판단했습니다.

soft delete는 도메인 계층이 담당하는 `CRUD`의 영역이므로, 테스트를 한다면 repository에서 테스트 코드를 작성해야 하지 않을까요?

현재는 hard delete 방식으로 삭제를 하고, 후순위로 soft delete를 구현하는 것에는 이견이 없습니다!

Hyyena reviewed (approved):
고생하셨습니다😊
